File: interfaces/api/v1/world/knowledge.py
"""Knowledge API routes"""
from fastapi import APIRouter, Depends, HTTPException, Query
from typing import Optional
from pydantic import BaseModel, Field

from application.world.services.knowledge_service import KnowledgeService
from application.world.dtos.knowledge_dto import (
    StoryKnowledgeDTO,
    ChapterSummaryDTO,
    KnowledgeTripleDTO,
    KnowledgeSearchResponseDTO
)
from interfaces.api.dependencies import get_knowledge_service
from domain.shared.exceptions import EntityNotFoundError


router = APIRouter(prefix="/novels/{novel_id}/knowledge", tags=["knowledge"])


# Request Models
class UpdateKnowledgeRequest(BaseModel):
    """更新知识图谱请求"""
    version: int = Field(default=1, description="数据版本")
    premise_lock: str = Field(default="", description="梗概锁定")
    chapters: list[ChapterSummaryDTO] = Field(default_factory=list, description="章节摘要列表")
    facts: list[KnowledgeTripleDTO] = Field(default_factory=list, description="知识三元组列表")


# Routes
@router.get("", response_model=StoryKnowledgeDTO)
async def get_knowledge(
    novel_id: str,
    service: KnowledgeService = Depends(get_knowledge_service)
):
    """获取知识图谱

    Args:
        novel_id: 小说ID
        service: 知识服务

    Returns:
        故事知识 DTO
    """
    knowledge = service.get_knowledge(novel_id)

    return StoryKnowledgeDTO(
        version=knowledge.version,
        premise_lock=knowledge.premise_lock,
        chapters=[
            ChapterSummaryDTO(
                chapter_id=ch.chapter_id,
                summary=ch.summary,
                key_events=ch.key_events,
                open_threads=ch.open_threads,
                consistency_note=ch.consistency_note,
                beat_sections=ch.beat_sections,
                sync_status=ch.sync_status
            )
            for ch in knowledge.chapters
        ],
        facts=[
            KnowledgeTripleDTO(
                id=fact.id,
                subject=fact.subject,
                predicate=fact.predicate,
                object=fact.object,
                chapter_id=fact.chapter_id,
                note=fact.note,
                entity_type=fact.entity_type,
                importance=fact.importance,
                location_type=fact.location_type,
                description=fact.description,
                first_appearance=fact.first_appearance,
                related_chapters=fact.related_chapters,
                tags=fact.tags,
                attributes=fact.attributes,
                confidence=fact.confidence,
                source_type=fact.source_type,
                subject_entity_id=fact.subject_entity_id,
                object_entity_id=fact.object_entity_id,
                provenance=list(getattr(fact, "provenance", []) or []),
            )
            for fact in knowledge.facts
        ]
    )


@router.put("", response_model=StoryKnowledgeDTO)
async def update_knowledge(
    novel_id: str,
    request: UpdateKnowledgeRequest,
    service: KnowledgeService = Depends(get_knowledge_service)
):
    """更新知识图谱

    Args:
        novel_id: 小说ID
        request: 更新请求
        service: 知识服务

    Returns:
        更新后的故事知识 DTO
    """
    import sys
    print(f"[API] update_knowledge called for {novel_id}, facts: {len(request.facts)}", file=sys.stderr, flush=True)

    try:
        data = {
            "version": request.version,
            "premise_lock": request.premise_lock,
            "chapters": [ch.model_dump() for ch in request.chapters],
            "facts": [fact.model_dump() for fact in request.facts]
        }

        print(f"[API] Calling service.update_knowledge", file=sys.stderr, flush=True)
        knowledge = service.update_knowledge(novel_id, data)
        print(f"[API] service.update_knowledge completed", file=sys.stderr, flush=True)
    except Exception as e:
        print(f"[API] Exception: {e}", file=sys.stderr, flush=True)
        import traceback
        traceback.print_exc()
        raise

    return StoryKnowledgeDTO(
        version=knowledge.version,
        premise_lock=knowledge.premise_lock,
        chapters=[
            ChapterSummaryDTO(
                chapter_id=ch.chapter_id,
                summary=ch.summary,
                key_events=ch.key_events,
                open_threads=ch.open_threads,
                consistency_note=ch.consistency_note,
                beat_sections=ch.beat_sections,
                sync_status=ch.sync_status
            )
            for ch in knowledge.chapters
        ],
        facts=[
            KnowledgeTripleDTO(
                id=fact.id,
                subject=fact.subject,
                predicate=fact.predicate,
                object=fact.object,
                chapter_id=fact.chapter_id,
                note=fact.note,
                entity_type=fact.entity_type,
                importance=fact.importance,
                location_type=fact.location_type,
                description=fact.description,
                first_appearance=fact.first_appearance,
                related_chapters=fact.related_chapters,
                tags=fact.tags,
                attributes=fact.attributes,
                confidence=fact.confidence,
                source_type=fact.source_type,
                subject_entity_id=fact.subject_entity_id,
                object_entity_id=fact.object_entity_id,
                provenance=list(getattr(fact, "provenance", []) or []),
            )
            for fact in knowledge.facts
        ]
    )


@router.get("/search", response_model=KnowledgeSearchResponseDTO)
async def search_knowledge(
    novel_id: str,
    q: str = Query(..., description="搜索查询"),
    k: int = Query(6, ge=1, le=50, description="返回结果数量"),
    service: KnowledgeService = Depends(get_knowledge_service)
):
    """搜索知识图谱

    Args:
        novel_id: 小说ID
        q: 搜索查询
        k: 返回结果数量
        service: 知识服务

    Returns:
        搜索结果
    """
    result = service.search_knowledge(novel_id, q, k)
    return KnowledgeSearchResponseDTO(**result)

File: frontend/src/api/knowledge.ts
import axios from 'axios'

const request = axios.create({
  baseURL: '/api/v1',
  timeout: 30000,
})

request.interceptors.response.use(response => response.data)

// TypeScript interfaces
export interface ChapterSummary {
  chapter_id: number
  summary: string
  key_events: string
  open_threads: string
  consistency_note: string
  beat_sections: string[]
  sync_status: string
}

export interface KnowledgeTriple {
  id: string
  subject: string
  predicate: string
  object: string
  chapter_id: number | null
  note: string
  entity_type?: 'character' | 'location'
  importance?: 'primary' | 'secondary' | 'minor' | 'core' | 'important' | 'normal'
  location_type?: 'city' | 'region' | 'building' | 'faction' | 'realm'
  description?: string
  first_appearance?: number
  related_chapters?: number[]
  tags?: string[]
  attributes?: Record<string, any>
  source_type?: string
  subject_entity_id?: string
  object_entity_id?: string
  /** 服务端推断溯源；PUT 时忽略 */
  provenance?: Array<{
    id?: string
    story_node_id?: string | null
    chapter_element_id?: string | null
    rule_id: string
    role?: string
  }>
}

export interface StoryKnowledge {
  version: number
  premise_lock: string
  chapters: ChapterSummary[]
  facts: KnowledgeTriple[]
}

export interface KnowledgeSearchHit {
  id: string
  text: string
  meta?: {
    type?: string
    id?: string
    [key: string]: any
  }
}

export interface KnowledgeSearchResponse {
  hits: KnowledgeSearchHit[]
}

export const knowledgeApi = {
  /**
   * Get knowledge graph for a novel
   */
  getKnowledge: (novelId: string) =>
    request.get(`/novels/${novelId}/knowledge`) as Promise<StoryKnowledge>,

  /**
   * Update knowledge graph for a novel
   */
  updateKnowledge: (novelId: string, data: StoryKnowledge) =>
    request.put(`/novels/${novelId}/knowledge`, data) as Promise<StoryKnowledge>,

  /** 与 updateKnowledge 相同（兼容旧组件名） */
  putKnowledge: (novelId: string, data: StoryKnowledge) =>
    request.put(`/novels/${novelId}/knowledge`, data) as Promise<StoryKnowledge>,

  /**
   * Search knowledge graph
   */
  searchKnowledge: (novelId: string, query: string, k = 6) =>
    request.get(`/novels/${novelId}/knowledge/search`, {
      params: { q: query, k }
    }) as Promise<KnowledgeSearchResponse>,

  /**
   * AI generate (or regenerate) initial Knowledge for a novel
   * POST /api/v1/novels/{novelId}/knowledge/generate
   */
  generateKnowledge: (novelId: string) =>
    request.post<{ success: boolean; message: string; facts_count: number; premise_lock: string }>(
      `/novels/${novelId}/knowledge/generate`,
      {},
      { timeout: 120_000 }
    ) as Promise<{ success: boolean; message: string; facts_count: number; premise_lock: string }>,
}
